plugins/GSdx/GSOsdManager.cpp

```diff
@@ -0,0 +1,403 @@
+			fprintf(stderr, "failed to load char '%c' aka %d\n", pair.first, pair.first);
+			continue;
+		}
+
+		// Size of char
+		pair.second.ax = m_face->glyph->advance.x >> 6;
+		pair.second.ay = m_face->glyph->advance.y >> 6;
+
+		pair.second.bw = m_face->glyph->bitmap.width;
+		pair.second.bh = m_face->glyph->bitmap.rows;
+
+		pair.second.bl = m_face->glyph->bitmap_left;
+		pair.second.bt = m_face->glyph->bitmap_top;
+
+		GSVector4i r(x, 0, x+pair.second.bw, pair.second.bh);
+		if (r.width())
+			t->Update(r, m_face->glyph->bitmap.buffer, m_face->glyph->bitmap.pitch);
+
+		if (r.width() > m_max_width) m_max_width = r.width();
+
+		pair.second.tx = (float)x / atlas_w;
+		pair.second.ty = (float)pair.second.bh / atlas_h;
+		pair.second.tw = (float)pair.second.bw / atlas_w;
+
+		x += pair.second.bw;
+	}
+
+	m_texture_dirty = false;
+
+#if 0
+	m_indicator["Up"].glyph = c_info['U'];
+	m_indicator["Down"].glyph = c_info['D'];
+	m_indicator["Left"].glyph = c_info['L'];
+	m_indicator["Right"].glyph = c_info['R'];
+	m_indicator["Triangle"].glyph = c_info['T'];
+	m_indicator["Cross"].glyph = c_info['X'];
+	m_indicator["Square"].glyph = c_info['S'];
+	m_indicator["Circle"].glyph = c_info['O'];
+#endif
+}
+
+/* This is dumb in that it doesn't check for malformed UTF8. This function
+ * is not expected to operate on user input, but only on compiled in strings */
+void dumb_utf8_to_utf32(const char *utf8, char32_t *utf32, unsigned size) {
+	while(*utf8 && --size) {
+		if((*utf8 & 0xF1) == 0xF0) {
+			*utf32++ = (utf8[0] & 0x07) << 18 | (utf8[1] & 0x3F) << 12 | (utf8[2] & 0x3F) << 6 | utf8[3] & 0x3F;
+			utf8 += 4;
+		} else if((*utf8 & 0xF0) == 0xE0) {
+			*utf32++ =                          (utf8[0] & 0x0F) << 12 | (utf8[1] & 0x3F) << 6 | utf8[2] & 0x3F;
+			utf8 += 3;
+		} else if((*utf8 & 0xE0) == 0xC0) {
+			*utf32++ =                                                   (utf8[0] & 0x1F) << 6 | utf8[1] & 0x3F;
+			utf8 += 2;
+		} else if((*utf8 & 0x80) == 0x00) {
+			*utf32++ =                                                                           utf8[0] & 0x7F;
+			utf8 += 1;
+		}
+	}
+
+	if(size) *utf32 = *utf8; // Copy NUL char
+}
+
+void GSOsdManager::AddGlyph(char32_t codepoint) {
+	if(c_info.count(codepoint) == 0) {
+		m_texture_dirty = true;
+		c_info[codepoint]; // add it
+		if(FT_HAS_KERNING(m_face)) {
+			FT_UInt new_glyph = FT_Get_Char_Index(m_face, codepoint);
+			for(auto pair : c_info) {
+				FT_Vector delta;
+
+				FT_UInt glyph_index = FT_Get_Char_Index(m_face, pair.first);
+				FT_Get_Kerning(m_face, glyph_index, new_glyph, FT_KERNING_DEFAULT, &delta);
+				k_info[std::make_pair(pair.first, codepoint)] = delta.x >> 6;
+			}
+		}
+	}
+}
+
+void GSOsdManager::Log(const char *utf8, uint32 color) {
+	if(!m_log_enabled)
+		return;
+
+	char32_t buffer[256];
+	dumb_utf8_to_utf32(utf8, buffer, countof(buffer));
+	m_log.push_back((log_info){color, buffer, std::chrono::system_clock::time_point()});
```

Comment:
Won't  `emplace_back` be a better API. This way you can create a log_info directly in m_log.